designs/use-selectors-at-cache.md

```diff
@@ -0,0 +1,123 @@
+Filter cache ListWatch using selectors
+===================
+## Motivation
+
+The controller-runtime subscribe to changes using a cache mechanism that get
+updated using kubernetes informers, those informers implement a `ListWatch`
+that specify the namespace and type specified at controller builder time.
+
+So the only possible filtering of that cache is namespace and resource type,
+this means that if some user is interested only on Reconcile some resource with
+specific fields it will still cache all the instances on that type increasing
+the memory and CPU load. This also get worse in case controller-runtime is
+deployed as a daemonset instead of a deployment since all the nodes with
+daemonset's pods will expect to receive notification of all the instances and
+that can end with scalation problems depending on the type or resource it's
+watching.
+
+As side effect of this proposal the request that does not match the filtered
+will not be returned, so users has to be warned that if this filtering
+is used they know it and they are careful with the request they are doing.
+
+As a side note this proposal sidesteps some of the long standing issues with
+other [solutions](https://github.com/kubernetes-sigs/controller-runtime/issues/244#issuecomment-451336202)
+like problematic interactions between controllers or the need for a complex
+refactorings by adding the possibility for users to define the ListWatch
+selectors but with an API that will be used by users that know what are the
+implications.
```

Comment:
```suggestion
This proposal sidesteps the issue of "How to we plug this into the cache-backed client so that users get feedback when they request something that is not matching the caches filter" by only implementing the filter logic in the cache
package. This allows advanced users to combine a filtered cache with the already existing `NewCacheFunc` option in the manager and cluster package, while simultaneously hiding it from newer users that might not be aware of the implications and the associated foot-shoot potential.
```

Reply:
Done